### DISCUSSION TITLE
How to detect a rewrite in getServerSideProps?

### DISCUSSION THREAD

**Discussion by markedwards** [vercel/next.js#45316] [Help] [answered]
I'm doing a rewrite in next.config.js like:
```
        {
          source: "/@:id",
          destination: "/foo/:id",
        },
```
I want to only allow the rewritten URLs, and return a 404 for direct requests to /foo/{id}.  Is it possible to detect the rewrite in `getServerSideProps`?  I see a `_nextDidRewrite` property in `Symbol(NextRequestMeta)`, but that seems like digging where I should not dig.

Any suggestions here?

**Comment by markedwards** [comment]
So far the only things I can think of to solve this are:
- catchall route
- custom server

Is there some simple way to only permit a page to be access via a rewrite?

**Comment by muhayli** [comment, marked as answer]
i would suggest you use the API middleware for this case since you're dealing with SSR

**Comment by icyJoseph** [comment]
While it is possible to do this with plain middleware, by just checking the incoming pathname... you are asking for a `GSSP` solution. I think this should work:

```tsx
import type { GetServerSidePropsContext } from "next";
import Link from "next/link";

const FooPage = () => (
  <div>
    Foo: <Link href="/">Home</Link>
  </div>
);

export const getServerSideProps = (ctx: GetServerSidePropsContext) => {
  if (!ctx.req.url) return { notFound: true };

  // explicitly compare Direct requests to this page
  if (ctx.resolvedUrl === ctx.req.url) {
    return { notFound: true };
  }

  // explicitly compare next/link, next/router requests
  // to this page, aka, JSON request
  if (ctx.req.url.includes(ctx.resolvedUrl)) {
    return { notFound: true };
  }

  return { props: {} };
};

export default FooPage;
```

For completeness sake, the middleware function:

```ts
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req: NextRequest) {
  const { pathname, href } = req.nextUrl;

  if (pathname.startsWith("/foo")) {
    req.nextUrl.pathname = "/_404"; // send it to oblivion
    return NextResponse.rewrite(req.nextUrl);
  }

  return NextResponse.next();
}
```

- Note that in Next 13.1+ it is possible to return a response directly, https://nextjs.org/blog/next-13-1#nextjs-advanced-middleware

This middleware intercepts /foo requests, and it can even be enhanced to only run for `/foo` via Matchers: https://nextjs.org/docs/advanced-features/middleware#matcher

```ts
export const config = {
  matcher: '/foo/:id*',
}
```

However, limiting your middleware, because of this use case feels like a bit much, and at the same time, using a middleware that runs on all requests, feels like a bit much too.

And of course this has to be included in `next.config.js` for all cases:

```js
export const config = {
  matcher: '/about/:path*',
}
```

**Comment by markedwards** [comment]
Thanks for the help.  Middleware is a fine solution, I don't have a particular need to involve `getServerSideProps` actually.

One slightly odd thing, if I use this:
```
export const config = {
  matcher: '/foo/:id*',
};
```
then the redirect gets triggered on `/@id` when going back to it in the browser.  So, I load `/@id`, everything is fine, I click a link to another page, then go back, and `req.nextUrl.pathname` is now `/foo/id`.  If I include `/@:id*` in the matcher, or omit the matcher altogether, this does not occur:
```
export const config = {
  matcher: [
    '/foo/:id*',
    '/@:id*',
  ],
};
```

**Reply by icyJoseph** [reply]
Ah perhaps go with my GSSP solution then, less moving parts.

**Reply by muhayli** [reply]
good, mark this as. answered then

**Reply by markedwards** [reply]
The GSSP solution actually does not work when navigating back to the page.  In that case, this condition is true and results in an incorrect 404:
```
ctx.req.url.includes(ctx.resolvedUrl)
```
I’m happy with implementing the solution in middleware though.
